code/controllers/subsystems/zlevels.dm

```diff
@@ -0,0 +1,106 @@
+	/// This list contains the z-level numbers which can be accessed via space travel and the percentile chances to get there.
+	var/list/accessible_z_levels = list()
+	/// Z-levels available to various consoles, such as the crew monitor. Defaults to station_levels if unset.
+	var/list/map_levels
+	/// The turf type used when generating floors between Z-levels at startup.
+	var/base_floor_type = /turf/simulated/floor/airless 
+	/// Replacement area, if a base_floor_type is generated. Leave blank to skip.
+	var/base_floor_area
+	/// A list of connected z-levels to avoid repeatedly rebuilding connections
+	var/list/connected_z_cache = list()
+
+/datum/controller/subsystem/zlevels/Initialize(start_timeofday)
+	for(var/z = 1 to world.maxz)
+		var/obj/abstract/level_data/level = levels_by_z["[z]"]
+		if(level)
+			level.setup_level_data()
+		else
+			PRINT_STACK_TRACE("Missing z-level data object for z["[z]"]!")
+		report_progress("z[z]: [get_level_name(z)]")
+	. = ..()
+
+/datum/controller/subsystem/zlevels/proc/get_level_name(var/z)
+	z = "[z]"
+	if(!z)
+		return "Unknown Sector"
+	var/obj/abstract/level_data/level = levels_by_z[z]
+	if(level?.name)
+		return level.name
+	var/obj/effect/overmap/overmap_entity = global.overmap_sectors[z]
+	if(overmap_entity?.name)
+		return overmap_entity.name
+	return "Sector #[z]"
+
+/datum/controller/subsystem/zlevels/proc/increment_world_z_size(var/new_level_type, var/defer_setup = FALSE)
+
+	world.maxz++
+	connected_z_cache.Cut()
+	if(SSzcopy.zlev_maximums.len)
+		SSzcopy.calculate_zstack_limits()
+
+	if(!new_level_type)
+		PRINT_STACK_TRACE("Missing z-level data type for z["[world.maxz]"]!")
+		return
+
+	var/obj/abstract/level_data/level = new new_level_type(locate(round(world.maxx*0.5), round(world.maxz*0.5), world.maxz), defer_setup)
+	if(level.base_turf_type && level.base_turf_type != world.turf)
+		for(var/turf/T as anything in block(locate(1, 1, .),locate(world.maxx, world.maxy, level.my_z)))
+			T.ChangeTurf(level.base_turf_type)
+	if(level.base_area_type && level.base_area_type != world.area)
+		var/area/A = new level.base_area_type
+		for(var/turf/T as anything in block(locate(1, 1, .),locate(world.maxx, world.maxy, level.my_z)))
+			ChangeArea(T, A)
+	return level
+
+/datum/controller/subsystem/zlevels/proc/levels_are_z_connected(var/za, var/zb)
+	return (za > 0 && zb > 0 && za <= world.maxz && zb <= world.maxz) && ((za == zb) || ((length(connected_z_cache) >= za && connected_z_cache[za] && length(connected_z_cache[za]) >= zb) ? connected_z_cache[za][zb] : are_connected_levels(za, zb)))
+
+/datum/controller/subsystem/zlevels/proc/get_connected_levels(z)
+	. = list(z)
+	// Traverse up and down to get the multiz stack.
+	for(var/level = z, HasBelow(level), level--)
+		. |= level-1
+	for(var/level = z, HasAbove(level), level++)
+		. |= level+1
+	// Check stack for any laterally connected neighbors.
+	for(var/tz in .)
+		var/obj/abstract/level_data/level = levels_by_z["[tz]"]
```

Comment:
I hope you dont want to use this in any hot code like atmos because string interp is agonizingly slow